### PR TITLE
Copy-on-write for BatchRequest

### PR DESCRIPTION
**Pull request by tbg**
This removes one location in the code in which we
modified data in the incoming request which could
possibly be used concurrently without our knowledge.

This change makes sure that we shallow-copy just
enough to rewrite into our own personal copies.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/4170)

### CODE REVIEW

File: kv/dist_sender.go [hunk 1]
```diff
@@ -442,16 +442,22 @@ func (ds *DistSender) Send(ctx context.Context, ba roachpb.BatchRequest) (*roach
...
 			//
-			// TODO(pmattis): We should shallow-copy the Txn.
-			ba.Txn.CertainNodes.Add(nDesc.NodeID)
+			// Taking care below to not modify any memory referenced from
+			// our BatchRequest which may be shared with others.
+			// First, get a shallow clone of our txn (since that holds the
+			// NodeList struct).
+			txnShallow := *ba.Txn
+			// Next, zero out the NodeList pointer. That makes sure that
+			// if we had something of size zero but with capacity, we don't
```
Comment (tamird) [inline kv/dist_sender.go]: how would that happen?

Comment (tschottdorf) [inline kv/dist_sender.go]: I don't know how, but doesn't seem like this code should make assumptions.

Comment (tamird) [inline kv/dist_sender.go]: Oh, it can happen if another shallow copy has called Add. duh

**Comment by tamird** [comment]
LGTM

File: kv/dist_sender.go (hunk 1 above)
Comment (tschottdorf) [inline kv/dist_sender.go]: not really, then that other shallow copy would have its own slice that we wouldn't know about. I suppose you'd have to maliciously create a `NodeList{Nodes: make([]NodeID, 0, 1)}` or something. Anyway, better safe than sloppy.

Comment (tamird) [inline kv/dist_sender.go]: if ba.Txn.Nodes is non-nil coming in, shallow copies will share the underlying array, and bad things can happen. isn't that true?

File: kv/dist_sender.go [hunk 2]
```diff
@@ -442,16 +442,22 @@ func (ds *DistSender) Send(ctx context.Context, ba roachpb.BatchRequest) (*roach
...
+			// re-use the existing space (which others may also use).
+			txnShallow.CertainNodes.Nodes = nil
```
Comment (petermattis) [inline kv/dist_sender.go]: Shouldn't you be copying `CertainNodes` here? This is a functionality change from the previous code. We were appending to `CertainNodes`, now we're clearing and assigning to it.

File: kv/dist_sender.go (hunk 1 above)
Comment (tschottdorf) [inline kv/dist_sender.go]: yes, but if we're here then it has length zero.

File: kv/dist_sender.go (hunk 2 above)
Comment (tschottdorf) [inline kv/dist_sender.go]: Easy to miss, but in this branch we know it's empty.

**Pull request merged**